src/xmlwriter.rs

```diff
@@ -63,15 +62,27 @@ impl XMLWriter {
     }
 
     // Write an XML start tag with attributes.
-    pub(crate) fn xml_start_tag_attr(&mut self, tag: &str, attributes: &Vec<(&str, String)>) {
-        let mut attribute_str = "".to_string();
+    pub(crate) fn xml_start_tag_attr(&mut self, tag: &str, attributes: &[(&str, String)]) {
+        let approx_len = tag.len() + 6 * attributes.len() + 2;
+        let mut tag_attr = String::with_capacity(approx_len);
+
+        write!(&mut tag_attr, "<{tag}",).unwrap();
 
         for attribute in attributes {
-            let pair = format!(r#" {}="{}""#, attribute.0, escape_attributes(&attribute.1));
-            attribute_str.push_str(&pair);
+            write!(
```

Comment:
We could just `write()` directly to the cursor there and above and remove the string allocation altogether. Something like this:

```rust
    pub(crate) fn xml_start_tag_attr(&mut self, tag: &str, attributes: &[(&str, String)]) {
        write!(&mut self.xmlfile, r"<{tag}").expect("Couldn't write to file");

        for attribute in attributes {
            write!(
                &mut self.xmlfile,
                r#" {}="{}""#,
                attribute.0,
                escape_attributes(&attribute.1)
            )
            .expect("Couldn't write to file");
        }

        write!(&mut self.xmlfile, ">").expect("Couldn't write to file");
    }
```

Reply:
That's even better. I¶m working on it.

Reply:
I just checked the C version of the library and this is basically what I did there. So I will probably make that change on main and you can rebase the other changes against it.

Reply:
Fixed already ;)

Reply:
I've already made the change and few other minor one. I'm just waiting for the tests to pass to push it to main so you will need to rebase.

Reply:
No problem.